packages/core/src/client.ts

```diff
@@ -151,89 +151,32 @@ export abstract class Client {
   }
 
   notify(noticeable: Noticeable, name: string | Partial<Notice> = undefined, extra: Partial<Notice> = undefined): boolean {
-    let preConditionError: Error = null
     const notice = this.makeNotice(noticeable, name, extra)
-    if (!notice) {
-      this.logger.debug('failed to build error report')
-      preConditionError = new Error('failed to build error report')
-    }
 
-    if (!preConditionError && this.config.reportData === false) {
-      this.logger.debug('skipping error report: honeybadger.js is disabled', notice)
-      preConditionError = new Error('honeybadger.js is disabled')
-    }
-
-    if (!preConditionError && this.__developmentMode()) {
-      this.logger.log('honeybadger.js is in development mode; the following error report will be sent in production.', notice)
-      preConditionError = new Error('honeybadger.js is in development mode')
-    }
-
-    if (!preConditionError && !this.config.apiKey) {
-      this.logger.warn('could not send error report: no API key has been configured', notice)
-      preConditionError = new Error('missing API key')
-    }
     // we need to have the source file data before the beforeNotifyHandlers,
     // in case they modify them
     const sourceCodeData = notice && notice.backtrace ? notice.backtrace.map(trace => shallowClone(trace) as BacktraceFrame) : null
-    const beforeNotifyResult = runBeforeNotifyHandlers(notice, this.__beforeNotifyHandlers)
-    if (!preConditionError && !beforeNotifyResult) {
-      this.logger.debug('skipping error report: beforeNotify handlers returned false', notice)
-      preConditionError = new Error('beforeNotify handlers returned false')
-    }
+    const preConditionsResult = this.__runPreconditions(notice)
+    if (preConditionsResult instanceof Error) {
+      runAfterNotifyHandlers(notice, this.__afterNotifyHandlers, preConditionsResult)
 
-    if (preConditionError) {
-      runAfterNotifyHandlers(notice, this.__afterNotifyHandlers, preConditionError)
       return false
     }
 
-    this.addBreadcrumb('Honeybadger Notice', {
-      category: 'notice',
-      metadata: {
-        message: notice.message,
-        name: notice.name,
-        stack: notice.stack
-      }
-    })
+    if (preConditionsResult instanceof Promise) {
+      preConditionsResult.then((result) => {
+        if (result instanceof Error) {
+          runAfterNotifyHandlers(notice, this.__afterNotifyHandlers, result)
 
-    const breadcrumbs = this.__store.getContents('breadcrumbs')
-    notice.__breadcrumbs = this.config.breadcrumbsEnabled ? breadcrumbs : []
+          return false
+        }
+        return this.__send(notice, sourceCodeData)
+      })
 
-    getSourceForBacktrace(sourceCodeData, this.__getSourceFileHandler)
-      .then(sourcePerTrace => {
-        sourcePerTrace.forEach((source, index) => {
-          notice.backtrace[index].source = source
-        })
+      return true
```

Comment:
In case of an async beforeNotify handler, we return true regardless of the result.

Reply:
I understand why this is working this way, but this is smelly. A user would definitely not intuitively know how to interpret this boolean. 

I took a quick look through the documentation but I couldn't find anywhere where we explain this return value or give an example of its use. Do we know if users rely on it at all? 

If the return value simply indicates that we ran the precondition checks (which are all sync) and does not provide any information about the results of the `beforeNotify` handlers, this could be understandable for a user. But imo the idea that it returns a meaningful result of the `beforeNotify` handlers, but only if all of them are sync is confusing. Getting `true` back when it actually means 'promises are pending'... I think it's worse than getting nothing back at all. 

I think I'd be in favor of just returning the result of the precondition checks and clearly documenting that behavior somewhere. I'm not sure if this would constitute a "breaking change" or not if the prior behavior wasn't documented... theres a chance that savvy users are relying on knowing if their `beforeNotify` handler failed even if its not in the docs.

Reply:
> I took a quick look through the documentation but I couldn't find anywhere where we explain this return value or give an example of its use. Do we know if users rely on it at all?

Knowing the answer to this question would greatly help us to decide whether we should be trying to stay backwards compatible or not. Actually, that's why it took us so long to work on this feature request. Unfortunately, we do not have this kind of information.

> If the return value simply indicates that we ran the precondition checks (which are all sync) and does not provide any information about the results of the beforeNotify handlers, this could be understandable for a user.

@joshuap explained that the boolean result can be used as an indicator to know whether the notice will be reported to HB (in a synchronous manner). I think it's a convention that came from other HB client integrations, but I'm not sure how much is used today.

> But imo the idea that it returns a meaningful result of the beforeNotify handlers, but only if all of them are sync is confusing.

True, I agree with you. Since we are missing usage data (how is the result of notify being used), we tried to implement the new feature (async beforeNotify) without breaking the existing functionality:
- the result of notify() remains consistent if sync beforeNotify handlers are used
- the result of notify() is always true if async beforeNotify handlers are used <- this is the new feature and we can document it to manage expectations

> Getting true back when it actually means 'promises are pending'...

You'd be getting that only with the new behavior, which we can document today.

> I think I'd be in favor of just returning the result of the precondition checks and clearly documenting that behavior somewhere. I'm not sure if this would constitute a "breaking change" or not if the prior behavior wasn't documented... theres a chance that savvy users are relying on knowing if their beforeNotify handler failed even if its not in the docs.

If we go this route, we'll have to make notify an async function, since we support async precondition checks, right? This goes against our approach of "fire-and-forget" for error reporting.

Btw, I'm loving the fact that you have the same concerns I had in the past when I first encountered this feature request 😄 .

Reply:
I added documentation about [Honeybadger.notify() return value](https://github.com/honeybadger-io/docs/pull/316/files#diff-8658674704fa4cb438d9f97f2ca0233be7cba8012c578766ebf88ecf20c48ea6R69).

Reply:
I sorta get what's going on here... sounds correct. :grin:

We used to use the return value pattern in our Ruby gem (and may still elsewhere, I forget), but at least for Ruby we've since moved to calling a method on the notice to prevent it from being sent:

https://docs.honeybadger.io/lib/ruby/getting-started/ignoring-errors/#ignore-programmatically